File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/internals/CharUtil.java
package toolgood.algorithm.internals;

import java.util.ArrayList;
import java.util.List;

public class CharUtil {
    public static char StandardChar(char c)
    {
        if (c <= 0) return c;
        char o = (char)c;
        if (o == '‘') return '\'';
        if (o == '’') return '\'';
        if (o == '“') return '"';
        if (o == '”') return '"';
        if (o == '〔') return '(';
        if (o == '〕') return ')';
        if (o == '＝') return '=';
        if (o == '＋') return '+';
        if (o == '－') return '-';
        if (o == '×') return '*';
        if (o == '÷') return '/';
        if (o == '／') return '/';

        if (c == 12288) {
            o = (char)32;
        } else if (c > 65280 && c < 65375) {
            o = (char)(c - 65248);
        }
        return Character.toUpperCase(o);
    }

    private static boolean EqualsOnce(String left, String right)
    {
        if (left.length() != right.length()) return false;
        for (int i = 0; i < left.length(); i++) {
            if (left.charAt(i)  != right.charAt(i)) {
                char a = StandardChar(left.charAt(i));
                char b = StandardChar(right.charAt(i));
                if (a != b) return false;
            }
        }
        return true;
    }

    public static boolean Equals(String left, String right)
    {
        if (left == null) return false;
        if (right == null) return false;
        return EqualsOnce(left, right);
    }
    public static boolean Equals(String left, String arg1, String arg2)
    {
        if (left == null) return false;
        if (arg1 != null && EqualsOnce(left, arg1))
            return true;
        if (arg2 != null && EqualsOnce(left, arg2))
            return true;
        return false;
    }

    public static boolean Equals(String left, String arg1, String arg2, String arg3)
    {
        if (left == null) return false;
        if (arg1 != null && EqualsOnce(left, arg1))
            return true;
        if (arg2 != null && EqualsOnce(left, arg2))
            return true;
        if (arg3 != null && EqualsOnce(left, arg3))
            return true;
        return false;
    }

    public static List<String> SplitFormula(String formula, List<Character> splitChars) {
        List<String> result = new ArrayList<>();
        boolean inSquareBrackets = false;
        boolean inBraceBrackets = false;
        int inBracketsCount = 0;
        boolean inText = false;
        char textChar = (char) 0;

        StringBuilder str = new StringBuilder();
        Integer i = 0;
        while (i < formula.length()) {
            char c = formula.charAt(i);
            if (inSquareBrackets) {
                str.append(c);
                if (c == ']') inSquareBrackets = false;
            } else if (inBraceBrackets) {
                str.append(c);
                if (c == '}') inBraceBrackets = false;
            } else if (inText) {
                str.append(c);
                if (c == '\\') {
                    i++;
                    if (i < formula.length()){
                        str.append(formula.charAt(i));
                    }
                } else if (c == textChar) {
                    inText = false;
                }
            } else if (splitChars.contains(c) && inBracketsCount == 0) {
                result.add(str.toString());
                result.add(((Character) c).toString());
                str = new StringBuilder();
            } else {
                str.append(c);
                if (c == '\'' || c == '"' || c == '`') {
                    textChar = c;
                    inText = true;
                } else if (c == '[') {
                    inSquareBrackets = true;
                } else if (c == '{') {
                    inBraceBrackets = true;
                } else if (c == '(') {
                    inBracketsCount++;
                } else if (c == ')') {
                    inBracketsCount--;
                }
            }
            i++;
        }
        if (str.length() > 0)
            result.add(str.toString());
        return result;
    }

    public static List<String> SplitFormulaForAnd(String formula)
    {
        List<String> result = new ArrayList<>();
        boolean inSquareBrackets = false;
        boolean inBraceBrackets = false;
        int inBracketsCount = 0;
        boolean inText = false;
        char textChar = (char) 0;

        StringBuilder str = new StringBuilder();
        Integer i = 0;
        while (i < formula.length()) {
            char c = formula.charAt(i);
            if (inSquareBrackets) {
                str.append(c);
                if (c == ']') inSquareBrackets = false;
            } else if (inBraceBrackets) {
                str.append(c);
                if (c == '}') inBraceBrackets = false;
            } else if (inText) {
                str.append(c);
                if (c == '\\') {
                    i++;
                    if (i < formula.length()){
                        str.append(formula.charAt(i));
                    }
                } else if (c == textChar) {
                    inText = false;
                }
            } else if (c == '&' && inBracketsCount == 0) {
                if (i + 1 < formula.length() && formula.charAt(i + 1) == '&') {
                    i++;
                    result.add(str.toString());
                    str = new StringBuilder();
                } else {
                    result.add(str.toString());
                }
            } else if (c == '|' && inBracketsCount == 0) {
                if (i + 1 < formula.length() && formula.charAt(i + 1) == '|') {
                    i++;
                    result.add(str.toString());
                    str = new StringBuilder();
                    str.append(String.join("&&", result));
                    str.append("||");
                    result.clear();
                } else {
                    result.add(str.toString());
                }
            } else {
                str.append(c);
                if (c == '\'' || c == '"' || c == '`') {
                    textChar = c;
                    inText = true;
                } else if (c == '[') {
                    inSquareBrackets = true;
                } else if (c == '{') {
                    inBraceBrackets = true;
                } else if (c == '(') {
                    inBracketsCount++;
                } else if (c == ')') {
                    inBracketsCount--;
                }
            }
            i++;
        }
        if (str.length() > 0)
            result.add(str.toString());
        return result;
    }

}

File: csharp/ToolGood.Algorithm.Fast/Internals/CharUtil.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace ToolGood.Algorithm.Internals
{
    internal static class CharUtil
    {
        public static char StandardChar(char o)
        {
            if (o <= 'Z') return o;
            if (o > 127) {
                if (o == '‘') return '\'';
                if (o == '’') return '\'';
                if (o == '“') return '"';
                if (o == '”') return '"';
                if (o == '〔') return '(';
                if (o == '〕') return ')';
                if (o == '＝') return '=';
                if (o == '＋') return '+';
                if (o == '－') return '-';
                if (o == '×') return '*';
                if (o == '÷') return '/';
                if (o == '／') return '/';
                //if (o == '【') return '[';
                //if (o == '】') return ']';
                //if (o == '（') return '(';
                //if (o == '）') return ')';
                //if (o == '，') return ',';
                //if (o == '！') return '!';
                if (o == 12288) return (char)32;
                if (o <= 65280) {
                    return o;
                } else if (o < 65375) {
                    o = (char)(o - 65248);
                }
            }
            return char.ToUpperInvariant(o);
        }

        public static string StandardString(string s)
        {
            var sb = new StringBuilder(s.Length);
            for (int i = 0; i < s.Length; i++) {
                sb.Append(StandardChar(s[i]));
            }
            return sb.ToString();
        }

        public static bool Equals(string left, char right)
        {
            if (left.Length != 1) return false;
            if (left[0] != right) {
                var a = StandardChar(left[0]);
                if (a != right) return false;
            }
            return true;
        }

        public static bool Equals(string left, string right)
        {
            if (left.Length != right.Length) return false;
            for (int i = 0; i < left.Length; i++) {
                if (left[i] != right[i]) {
                    var a = StandardChar(left[i]);
                    if (a != right[i]) return false;
                }
            }
            return true;
        }

        public static bool Equals(string left, string arg1, string arg2)
        {
            if (Equals(left, arg1)) return true;
            if (Equals(left, arg2)) return true;
            return false;
        }

        public static bool Equals(string left, string arg1, string arg2, string arg3)
        {
            if (Equals(left, arg1)) return true;
            if (Equals(left, arg2)) return true;
            if (Equals(left, arg3)) return true;
            return false;
        }

        public static List<String> SplitFormula(String formula, List<char> splitChars)
        {
            List<String> result = new List<String>();
            bool inSquareBrackets = false;
            bool inBraceBrackets = false;
            int inBracketsCount = 0;
            bool inText = false;
            char textChar = (char)0;

            StringBuilder str = new StringBuilder();
            int i = 0;
            while (i < formula.Length) {
                char c = formula[i];
                if (inSquareBrackets) {
                    str.Append(c);
                    if (c == ']') inSquareBrackets = false;
                } else if (inBraceBrackets) {
                    str.Append(c);
                    if (c == '}') inBraceBrackets = false;
                } else if (inText) {
                    str.Append(c);
                    if (c == '\\') {
                        i++;
                        if (i < formula.Length) {
                            str.Append(formula[i]);
                        }
                    } else if (c == textChar) {
                        inText = false;
                    }
                } else if (splitChars.Contains(c) && inBracketsCount == 0) {
                    result.Add(str.ToString());
                    result.Add(c.ToString());
                    str = new StringBuilder();
                } else {
                    str.Append(c);
                    if (c == '\'' || c == '"' || c == '`') {
                        textChar = c;
                        inText = true;
                    } else if (c == '[') {
                        inSquareBrackets = true;
                    } else if (c == '{') {
                        inBraceBrackets = true;
                    } else if (c == '(') {
                        inBracketsCount++;
                    } else if (c == ')') {
                        inBracketsCount--;
                    }
                }
                i++;
            }
            if (str.Length > 0)
                result.Add(str.ToString());
            return result;
        }
    }
}